src/upsert.rs

```diff
@@ -68,60 +72,79 @@ pub trait FindexUpsert<
-                additions.len() + deletions.len(),
-            );
+        // a reverse map of Keyword UIDs (e.g. hashes) to Keywords
+        let mut uid_to_keyword = HashMap::<Uid<UID_LENGTH>, Keyword>::new();
+
+        let mut new_chains = HashMap::<Keyword, HashMap<IndexedValue, BlockType>>::with_capacity(
+            additions.len() + deletions.len(),
+        );
         for (indexed_value, keywords) in additions {
             for keyword in keywords {
                 new_chains
-                    .entry(keyword.hash())
+                    .entry(keyword)
                     .or_default()
                     .insert(indexed_value.clone(), BlockType::Addition);
             }
         }
         for (indexed_value, keywords) in deletions {
             for keyword in keywords {
                 new_chains
-                    .entry(keyword.hash())
+                    .entry(keyword)
                     .or_default()
                     .insert(indexed_value.clone(), BlockType::Deletion);
             }
         }
         // Compute the Entry Table UIDs.
         let mut new_chains = new_chains
             .into_iter()
-            .map(|(keyword_hash, indexed_values)| {
-                (
-                    EntryTable::<UID_LENGTH, KWI_LENGTH>::generate_uid::<KMAC_KEY_LENGTH>(
-                        &k_uid,
-                        &keyword_hash,
-                        label,
-                    ),
-                    (keyword_hash, indexed_values),
-                )
+            .map(|(keyword, indexed_values)| {
+                let keyword_hash = keyword.hash();
+                let uid = EntryTable::<UID_LENGTH, KWI_LENGTH>::generate_uid::<KMAC_KEY_LENGTH>(
+                    &k_uid,
+                    &keyword_hash,
+                    label,
+                );
+                uid_to_keyword.insert(uid, keyword);
+                (uid, (keyword_hash, indexed_values))
             })
             .collect::<HashMap<_, _>>();
 
         // Query the Entry Table for these UIDs.
-        let mut encrypted_entry_table = self
+        let mut encrypted_entry_table: EncryptedTable<UID_LENGTH> = self
             .fetch_entry_table(Uids(new_chains.keys().copied().collect()))
             .await?
             .try_into()?;
 
+        // compute the map of keywords to booleans indicating whether the keyword
+        // was already present in the database
+        let mut keyword_presence = HashMap::new();
+        for uid in encrypted_entry_table.keys() {
+            if let Some(keyword) = uid_to_keyword.remove(uid) {
+                keyword_presence.insert(keyword, true);
+            }
+        }
+        // whatever is left in `uid_to_keyword` is not in the database,
+        // i.e. not already present. Update the keyword_presence accordingly
+        for (_, keyword) in uid_to_keyword {
+            keyword_presence.insert(keyword, false);
+        }
```

Comment:
This is not useful since you can remove a set instead of a map keyword <-> boolean.